src/hipscat/catalog/pixel_tree.py

```diff
@@ -0,0 +1,117 @@
+
+    There are a number of methods in this class which allow for quickly navigating through the
+    tree and performing operations to filter the pixels in the catalog.
+
+    Attributes: pixels: Nested dictionary of pixel nodes stored in the tree. Indexed by HEALPix
+    order then pixel number root_pixel: Root node of the tree. It's children are a subset of the
+    12 base HEALPix pixels
+    """
+
+    METADATA_ORDER_COLUMN_NAME = "order"
+    METADATA_PIXEL_COLUMN_NAME = "pixel"
+
+    def __init__(self, partition_info_df: pd.DataFrame):
+        """Initialises the Tree from the partition info metadata
+
+        Args:
+            partition_info_df: Partition Info metadata file loaded into a pandas DataFrame
+
+        Raises:
+            ValueError: Incorrectly configured pixel structure in metadata file
+        """
+        self.root_pixel = PixelNode(-1, -1, PixelNodeType.ROOT, None)
+        self.pixels = {-1: {-1: self.root_pixel}}
+        self._create_tree(partition_info_df)
+
+    def __len__(self):
+        pixel_count = 0
+        for _, order_pixels in self.pixels.items():
+            pixel_count += len(order_pixels)
+        return pixel_count
+
+    def contains(self, hp_order: int, hp_pixel: int) -> bool:
+        """Check if tree contains a node at a given order and pixel
+
+        Args:
+            hp_order: HEALPix order to check
+            hp_pixel: HEALPix pixel number to check
+
+        Returns:
+            True if the tree contains the pixel, False if not
+        """
+        return hp_order in self.pixels and hp_pixel in self.pixels[hp_order]
+
+    def get_node(self, hp_order: int, hp_pixel: int) -> PixelNode | None:
+        """Get the node at a given order and pixel
+
+        Args:
+            hp_order: HEALPix order to get
+            hp_pixel: HEALPix pixel number to get
+
+        Returns:
+            The PixelNode at the index, or None if a node does not exist
+        """
+        if self.contains(hp_order, hp_pixel):
+            return self.pixels[hp_order][hp_pixel]
+        return None
+
+    def _create_tree(self, partition_info_df: pd.DataFrame):
+        """Creates the tree by recursively creating parent nodes until reaching the root from
+        each leaf pixel
+
+        Validates the pixel structure as it does so by checking for duplicate pixels or pixels
+        defined at multiple orders
+        """
+        for _, row in partition_info_df.iterrows():
+            self._create_node_and_parent_if_not_exist(
+                row[self.METADATA_ORDER_COLUMN_NAME],
+                row[self.METADATA_PIXEL_COLUMN_NAME],
+                PixelNodeType.LEAF,
+            )
+
+    def _create_node_and_parent_if_not_exist(
+        self, hp_order: int, hp_pixel: int, node_type: PixelNodeType
+    ):
+        """Creates a node and adds to `self.pixels` in the tree, and recursively creates parent
+        node if parent does not exist"""
```

Comment:
The comment should include information on the arguments.

Reply:
Added, thanks